crates/ruff_python_formatter/src/pattern/pattern_match_mapping.rs

```diff
@@ -1,19 +1,197 @@
+use ruff_python_ast::node::AnyNodeRef;
+use ruff_python_ast::{Expr, Identifier, PatternMatchMapping};
+use ruff_python_ast::{Pattern, Ranged};
+use ruff_python_trivia::{SimpleTokenKind, SimpleTokenizer};
+use ruff_text_size::TextRange;
 
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
+use crate::comments::{leading_comments, trailing_comments, SourceComment};
+use crate::expression::parentheses::{
+    empty_parenthesized, parenthesized, NeedsParentheses, OptionalParentheses,
+};
+use crate::prelude::*;
+use crate::FormatNodeRule;
 
 #[derive(Default)]
 pub struct FormatPatternMatchMapping;
 
 impl FormatNodeRule<PatternMatchMapping> for FormatPatternMatchMapping {
     fn fmt_fields(&self, item: &PatternMatchMapping, f: &mut PyFormatter) -> FormatResult<()> {
+        let PatternMatchMapping {
+            keys,
+            patterns,
+            rest,
+            range: _,
+        } = item;
+
+        debug_assert_eq!(keys.len(), patterns.len());
+
+        let comments = f.context().comments().clone();
+        let dangling = comments.dangling(item);
+
+        if keys.is_empty() && rest.is_none() {
+            return empty_parenthesized("{", dangling, "}").fmt(f);
+        }
+
+        // This node supports three kinds of dangling comments. Most of the complexity originates
+        // with the rest pattern (`{**rest}`), since we can have comments around the `**`, but
+        // also, the `**rest` itself is not a node (it's an identifier), so comments that trail it
+        // are _also_ dangling.
+        //
+        // Specifically, we have these three sources of dangling comments:
+        // ```python
+        // {  # "open parenthesis comment"
+        //    key: pattern,
+        //    **  # end-of-line "double star comment"
+        //    # own-line "double star comment"
+        //    rest  # end-of-line "after rest comment"
+        //    # own-line "after rest comment"
+        // }
+        // ```
+        let (open_parenthesis_comments, double_star_comments, after_rest_comments) =
+            find_double_star(item, f.context().source()).map_or(
+                (dangling, dangling, dangling),
+                |(double_star, rest)| {
+                    let (open_parenthesis_comments, dangling) =
+                        dangling.split_at(dangling.partition_point(|comment| {
+                            comment.line_position().is_end_of_line()
+                                && comment.start() < double_star.start()
+                        }));
+                    let (double_star_comments, after_rest_comments) = dangling.split_at(
+                        dangling.partition_point(|comment| comment.start() < rest.start()),
+                    );
+                    (
+                        open_parenthesis_comments,
+                        double_star_comments,
+                        after_rest_comments,
+                    )
+                },
+            );
```

Comment:
Nit: I prefer to use `if_else` to reduce the nesting (or else, but our pedantic rules don't allow me to)

```suggestion
        let (open_parenthesis_comments, double_star_comments, after_rest_comments) = if let Some((double_star, rest)) = 
            find_double_star(item, f.context().source()) {
            	let (open_parenthesis_comments, dangling) =
                        dangling.split_at(dangling.partition_point(|comment| {
                            comment.line_position().is_end_of_line()
                                && comment.start() < double_star.start()
                        }));
                    let (double_star_comments, after_rest_comments) = dangling.split_at(
                        dangling.partition_point(|comment| comment.start() < rest.start()),
                    );
                    (
                        open_parenthesis_comments,
                        double_star_comments,
                        after_rest_comments,
                    )
            } else {
            	(&[], &[], &[])
          	};
```